api/grpcserver/node_service.go

```diff
@@ -0,0 +1,181 @@
+func NewNodeService(
+	net api.NetworkAPI, tx api.TxAPI, genTime api.GenesisTimeAPI, syncer api.Syncer) *NodeService {
+	return &NodeService{
+		Mesh:        tx,
+		GenTime:     genTime,
+		PeerCounter: peers.NewPeers(net, log.NewDefault("grpc_server.NodeService")),
+		Syncer:      syncer,
+	}
+}
+
+// Echo returns the response for an echo api request. It's used for E2E tests.
+func (s NodeService) Echo(ctx context.Context, in *pb.EchoRequest) (*pb.EchoResponse, error) {
+	log.Info("GRPC NodeService.Echo")
+	if in.Msg != nil {
+		return &pb.EchoResponse{Msg: &pb.SimpleString{Value: in.Msg.Value}}, nil
+	}
+	return nil, status.Errorf(codes.InvalidArgument, "Must include `Msg`")
+}
+
+// Version returns the version of the node software as a semver string
+func (s NodeService) Version(ctx context.Context, in *empty.Empty) (*pb.VersionResponse, error) {
+	log.Info("GRPC NodeService.Version")
+	return &pb.VersionResponse{
+		VersionString: &pb.SimpleString{Value: cmd.Version},
+	}, nil
+}
+
+// Build returns the build of the node software
+func (s NodeService) Build(ctx context.Context, in *empty.Empty) (*pb.BuildResponse, error) {
+	log.Info("GRPC NodeService.Build")
+	return &pb.BuildResponse{
+		BuildString: &pb.SimpleString{Value: cmd.Commit},
+	}, nil
+}
+
+// Status returns a status object providing information about the connected peers, sync status,
+// current and verified layer
+func (s NodeService) Status(ctx context.Context, request *pb.StatusRequest) (*pb.StatusResponse, error) {
+	log.Info("GRPC NodeService.Status")
+	return &pb.StatusResponse{
+		Status: &pb.NodeStatus{
+			ConnectedPeers: s.PeerCounter.PeerCount(),            // number of connected peers
+			IsSynced:       s.Syncer.IsSynced(),                  // whether the node is synced
+			SyncedLayer:    s.Mesh.LatestLayer().Uint64(),        // latest layer we saw from the network
+			TopLayer:       s.GenTime.GetCurrentLayer().Uint64(), // current layer, based on time
+			VerifiedLayer:  s.Mesh.LatestLayerInState().Uint64(), // latest verified layer
+		},
+	}, nil
+}
+
+// SyncStart requests that the node start syncing the mesh (if it isn't already syncing)
+func (s NodeService) SyncStart(ctx context.Context, request *pb.SyncStartRequest) (*pb.SyncStartResponse, error) {
+	log.Info("GRPC NodeService.SyncStart")
+	s.Syncer.Start()
+	return &pb.SyncStartResponse{
+		Status: &rpcstatus.Status{Code: int32(code.Code_OK)},
+	}, nil
+}
+
+// Shutdown requests a graceful shutdown
+func (s NodeService) Shutdown(ctx context.Context, request *pb.ShutdownRequest) (*pb.ShutdownResponse, error) {
+	log.Info("GRPC NodeService.Shutdown")
+	cmd.Cancel()
+	return &pb.ShutdownResponse{
+		Status: &rpcstatus.Status{Code: int32(code.Code_OK)},
+	}, nil
+}
+
+// STREAMS
+
+// StatusStream exposes a stream of node status updates
+func (s NodeService) StatusStream(request *pb.StatusStreamRequest, stream pb.NodeService_StatusStreamServer) error {
```

Comment:
should status be a stream or query?

Reply:
It's designed to be a stream. Queries are for historical data (which sort of doesn't make sense for status). With a stream, the client can always know immediately when the realtime status changes.